nexus/src/app/quiesce.rs

```diff
@@ -49,56 +31,163 @@ impl super::Nexus {
         opctx: &OpContext,
     ) -> LookupResult<QuiesceStatus> {
         opctx.authorize(authz::Action::Read, &authz::QUIESCE_STATE).await?;
-        let state = self.quiesce.borrow().clone();
-        let sagas_pending = self.saga_quiesce.sagas_pending();
+        let state = self.quiesce.state();
+        let sagas_pending = self.quiesce.sagas().sagas_pending();
         let db_claims = self.datastore().claims_held();
         Ok(QuiesceStatus { state, sagas_pending, db_claims })
     }
 }
 
-async fn do_quiesce(
-    quiesce: watch::Sender<QuiesceState>,
-    saga_quiesce: SagaQuiesceHandle,
+/// Describes the configuration and state around quiescing Nexus
+#[derive(Clone)]
+pub struct NexusQuiesceHandle {
+    log: Logger,
     datastore: Arc<DataStore>,
-) {
-    assert_matches!(*quiesce.borrow(), QuiesceState::WaitingForSagas { .. });
-    saga_quiesce.quiesce();
-    saga_quiesce.wait_for_quiesced().await;
-    quiesce.send_modify(|q| {
-        let QuiesceState::WaitingForSagas {
+    sagas: SagaQuiesceHandle,
+    state: watch::Sender<QuiesceState>,
+}
+
+impl NexusQuiesceHandle {
+    pub fn new(log: &Logger, datastore: Arc<DataStore>) -> NexusQuiesceHandle {
+        let my_log = log.new(o!("component" => "NexusQuiesceHandle"));
+        let saga_quiesce_log = log.new(o!("component" => "SagaQuiesceHandle"));
+        let sagas = SagaQuiesceHandle::new(saga_quiesce_log);
+        let (state, _) = watch::channel(QuiesceState::Undetermined);
+        NexusQuiesceHandle { log: my_log, datastore, sagas, state }
+    }
+
+    pub fn sagas(&self) -> SagaQuiesceHandle {
+        self.sagas.clone()
+    }
+
+    pub fn state(&self) -> QuiesceState {
+        self.state.borrow().clone()
+    }
+
+    pub fn set_quiescing(&self, quiescing: bool) {
+        let new_state = if quiescing {
+            let time_requested = Utc::now();
+            let time_draining_sagas = Instant::now();
+            QuiesceState::DrainingSagas { time_requested, time_draining_sagas }
+        } else {
+            QuiesceState::Running
+        };
+
+        let changed = self.state.send_if_modified(|q| {
+            match q {
+                QuiesceState::Undetermined => {
+                    info!(&self.log, "initial state"; "state" => ?new_state);
+                    *q = new_state;
+                    true
+                }
+                QuiesceState::Running if quiescing => {
+                    info!(&self.log, "quiesce starting");
+                    *q = new_state;
+                    true
+                }
+                _ => {
```

Comment:
Can we expand `_` to list all the cases? I find it hard to reason about `// All other cases are ...` when they aren't listed.